autogluon/utils/tabular/features/generators/abstract.py

```diff
@@ -0,0 +1,520 @@
+            y.index must be equal to X.index to avoid misalignment.
+        feature_metadata_in : FeatureMetadata, optional
+            Identical to providing feature_metadata_in during generator initialization. Ignored if self.feature_metadata_in is already specified.
+            If neither are set, feature_metadata_in will be inferred from the _infer_feature_metadata_in method.
+        **kwargs
+            Any additional arguments that a particular generator implementation could use. Passed to _fit_transform and _fit_generators methods.
+
+        Returns
+        -------
+        X_out : DataFrame object which is the transformed version of the input data X.
+
+        """
+        start_time = time.time()
+        self._log(20, f'Fitting {self.__class__.__name__}...')
+        if self._is_fit:
+            raise AssertionError(f'{self.__class__.__name__} is already fit.')
+        if y is not None and isinstance(y, Series):
+            if list(y.index) != list(X.index):
+                raise AssertionError(f'y.index and X.index must be equal when fitting {self.__class__.__name__}, but they differ.')
+
+        if self.reset_index:
+            X_index = copy.deepcopy(X.index)
+            X = X.reset_index(drop=True)  # TODO: Theoretically inplace=True avoids data copy, but can lead to altering of original DataFrame outside of method context.
+            if y is not None and isinstance(y, Series):
+                y = y.reset_index(drop=True)  # TODO: this assumes y and X had matching indices prior
+        else:
+            X_index = None
+        if self.column_names_as_str:
+            columns_orig = list(X.columns)
+            X.columns = X.columns.astype(str)  # Ensure all column names are strings
+            columns_new = list(X.columns)
+            if columns_orig != columns_new:
+                rename_map = {orig: new for orig, new in zip(columns_orig, columns_new)}
+                if feature_metadata_in is not None:
+                    feature_metadata_in.rename_features(rename_map=rename_map)
+                self._rename_features_in(rename_map)
+            else:
+                self.column_names_as_str = False  # Columns were already string, so don't do conversion. Better to error if they change types at inference.
+        self._infer_features_in_full(X=X, feature_metadata_in=feature_metadata_in)
+        if self.pre_drop_useless:
+            self._useless_features_in = self._get_useless_features(X)
+            if self._useless_features_in:
+                self._remove_features_in(self._useless_features_in)
+        if self.pre_enforce_types:
+            from .astype import AsTypeFeatureGenerator
+            self._pre_astype_generator = AsTypeFeatureGenerator(features_in=self.features_in, feature_metadata_in=self.feature_metadata_in, log_prefix=self.log_prefix + '\t')
+            self._pre_astype_generator.fit(X)
+
```

Comment:
Please add check for duplicated columns: this can happen in a few cases and might cause issues downstream: 
1) dataframe concatenation (on axis=1)
2) edge case with numeric and str column in passed dataframe:
```python
df = pd.DataFrame({'0': [1,2,3,4,5],0: [1,2,3,4,5],})
df.columns = df.columns.astype(str)
```

Reply:
Good point, added check and test